### ISSUE TITLE
Rethink screen wakelock - let the screen sleep in "normal" mode, but keep it on in "navigation" mode

### ISSUE THREAD

**Issue by denilsonsa** [organicmaps/organicmaps#1578]
#437 and #196 and 470bc6f231bcb93a8024e8407d5b7903affdd9d2 added a new setting "Allow screen to sleep". This setting is great for saving battery. However, sometimes that's desirable, sometimes not.

My problem: I tried using Organic Maps (Android) for a quick navigation (driving from point A to point B). However, having the screen to turn of while driving is less than ideal. In fact, it's dangerous. But at the same time I like letting it turn off automatically when idle, because I'm used to the behavior of the phone going to sleep on most apps.

---

**What happens:** There is one boolean toggle to let the screen turn off automatically, or to keep it always on. This setting is global, applies to the entire app, disregarding the context.

**What should happen:** I'm not sure exactly. Let me enumerate certain use-cases (AKA brainstorm):

* Thinking about the current map mode:
  * While idle in "normal" mode, the screen should turn **OFF** to save battery.
  * During "navigation" mode…
    * …while _driving_, most likely the screen should stay **ON**.
    * …while _cycling_ or _walking_, the screen could turn **OFF** automatically. Or maybe not. Let it stay **ON**, and let the user press the power button to manually turn it off.
* Thinking about charging state:
  * If the phone is being charged while the app is in use, most likely the user wants the screen to stay on.
  * If the phone is not charging, most likely the screen should go off to save the phone battery.
* Thinking about what the user is doing:
  * If the user is driving…
    * …most likely the user has the phone connected to a charger in the vehicle, so battery life is not an issue.
    * …most likely the screen timeout will happen after the vehicle is already on the move. (Because the user will start the navigation and then start driving.)
    * …then the user must keep the hands on the wheel and the eyes on the road. Any extra effort to turn on the screen and potentially unlock the phone can be dangerous.
  * If the user is cycling or walking…
    * …most likely the user is on battery, with or without an external power bank.
    * …then the user is moving slower (up to 20km/h), which is less dangerous than driving, and gives more time to react. It also means the user position updates slower, which means the user doesn't need to look at the screen as often.
    * …then most likely the user can reach the phone to turn it on or off. Although doing so while cycling can be cumbersome.
* Thinking about other app(s):
  * Google maps in normal mode…
      * …will turn the screen off when idle, just as any other app. (In other words, it won't acquire any screen wakelock.)
  * Google maps in navigation mode…
    * …will keep the screen on.
    * …will show up even when the phone is locked. (How to reproduce: enter navigation, then press the power button to turn off the screen and lock the phone, then press the power button again to wake up the phone → the Google Maps Navigation will show up, although the phone is still locked.)

---

In conclusion… What should Organic Maps do? I think it should treat the two modes in different ways. I like the way Google Maps behaves, mostly because that's what I'm most used to, but also because it seems to behave in a sane way on the different use-cases. But I also like the ability to tweak the settings to the user needs, and I don't want to disrupt any current use-cases. I'm also not sure how to empower the user to tweak the behavior without becoming a UX nightmare.

**Comment by biodranik** [comment]
What is the behavior of other apps?

**Comment by denilsonsa** [comment]
## [Google Maps](https://play.google.com/store/apps/details?id=com.google.android.apps.maps)

* At the "normal" view of the map…
    * …the screen turns off automatically according to the system settings.
* At the "directions" mode…
    * …the screen stays always on.
    * …the Google Maps in driving mode will show up instead (or before) the lock screen. (How to reproduce: press the power button to lock the phone, and then press it again to turn the screen on.)
    * …switching to another app will keep the driving map visible in a "[picture-in-picture](https://developer.android.com/guide/topics/ui/picture-in-picture)" style.
* At the "directions" "picture-in-picture" mode…
    * …the screen stays always on.
    * …locking the phone will NOT show up Google Maps before the lock screen. Instead, it just shows the standard notification, which is very minimal and only shows the next step of the directions.

## [Waze](https://play.google.com/store/apps/details?id=com.waze)

Waze is focused on driving, it doesn't support any other transportation method (no walking, no cycling, no public transportation). It supports either driving a private car, a taxi, or riding a motorcycle.

* At the "normal" mode…
    * …the screen stays always on. Given it's an app designed to be a company to the driver, regardless of the need for directions, it kinda makes sense. The driver leaves the phone mounted to the car, plugged to a charger, and with Waze open.
    * …, when in the background or at the lock screen: it shows a useless notification that Waze is still running. That's all. No directions in the notification.
* At the "directions" mode…
    * …it behaves just like the "normal" mode.

Waze also has an interesting setting:

![Screenshot of the Waze "Battery saver" settings screen](https://user-images.githubusercontent.com/121676/142757876-766137aa-84b8-428a-94e3-50e91c5f20cf.jpg)

> Waze will dim your screen when your battery is low and there are no upcoming turns. For full brightness, tap the screen.

This sounds like a great use-case for [`SCREEN_DIM_WAKE_LOCK`](https://stackoverflow.com/q/14226426). If there are no upcoming turns, let the screen dim, but allow the user to get normal brightness by tapping, and get normal brightness automatically near any turn. I'd prefer to have it enabled regardless of the battery level, because that will prevent the battery from getting too low.

## [Citymapper](https://play.google.com/store/apps/details?id=com.citymapper.app.release)

Citymapper is focused on basically anything but driving. It supports walking and public transportation, as well as cycling. It also shows information from other services to rent/share different kinds of vehicles (bikes, mopeds, cars) and even from Uber.

* At the "normal" mode…
    * …the screen turns off automatically according to the system settings.
* At the "directions" mode…
    * …the screen stays on.
    * …it shows a helpful notification showing the next step, as well as two arrows to view the following steps.
    * …at the lock screen, there are no fancy features: just the notification.
    * …I believe it will vibrate when it is time to take of the bus/train/metro/tram at the next stop. It's been a while since I used it, so I don't remember. I also don't remember if it vibrates when it is time to make a turn (useful if the phone is in your pocket while walking or cycling).

## [A Better Routeplanner (ABRP)](https://play.google.com/store/apps/details?id=com.iternio.abrpapp)

ABRP is focused on planning routes for EVs (electric vehicles), taking into account the stops for charging stations. The app feels like a webview, the interface doesn't always behave as users expect.

* At the "normal" mode…
    * …the screen turns off automatically according to the system settings.
* At the "directions" mode…
    * …the screen stays on.
    * …there is no notification.
    * …there is an option to open the route in Google Maps, which has more features and probably a better interface.

---

These are all the apps I have used in the past couple of years. For sure there are plenty more apps available, but I can't say about them. In summary:

| App | normal | directions | notification | extras |
|:--- |:------ |:---------- |:------------ |:------ |
| Google Maps | auto-off | stays on | next turn | PiP + shows before the lock screen |
| Waze | stays on | stays on | useless | optional screen dimming on low battery |
| Citymapper | auto-off | stays on | view multiple next steps | vibration |
| ABRP | auto-off | stays on | none | calculates EV battery level and routes through charging points |

**Comment by Der-Flip** [comment]
(feature idea) just because it was mentioned here:
Notifications would be quite a nice feature. When cycling I don't always want to have my phone in my pockets, but rather in my bicycle bag. Having notifications (maybe as an enableable toggle in the settings?) I could have these forwarded to my mi watch lite2 (or any other smart watch), thus not have the feeling of crushing my phone with my thigh / my phone crushing my thigh and stomach while I have it force-fit into my pocket...

**Comment by rancidfrog** [comment]
I am having a problem that if I force screen off with power button it causes blank screen on awake. And i have to close organic maps from recents, as it blocks on a blank screen.
I need to swipe notification bar and click settings to be able to unblock the phone then close app from recents

**Comment by biodranik** [comment]
@rancidfrog Did you try to turn off OM on the Lock Screen in OM's settings?

**Comment by rtsisyk** [comment]
The proposed behavior is implemented by #6053.

**Comment by denilsonsa** [comment]
Thank you, @rtsisyk!

After the screen wake lock feature is merged, it might be worth splitting this issue here into multiple: one for having a nice PiP (Picture-in-picture) mode, another for having useful turn-by-turn notifications (which may even show up on smart-watches, because wearable usually show notifications to the wearer), another for dimming the screen when there are no upcoming turns (replicating the Waze behavior).

**Comment by rtsisyk** [comment]
> another for having useful turn-by-turn notifications (which may even show up on smart-watches,

We have them since the last release, please check. 

> PiP (Picture-in-picture) mode

PIP is not really planned. Why do you need it?

**Comment by rtsisyk** [comment]
> I haven't tried "directions" on Organic Maps recently, so I missed the notification feature. Thanks for telling me, and sorry for the noise. :)

Thanks for the detailed analysis! I really appreciate your efforts to compare popular navigation apps.
Let's split this ticket:

- [Turn Notifications](https://github.com/organicmaps/organicmaps/pull/4611) - implemented
- [Proper Wake Lock](https://github.com/organicmaps/organicmaps/pull/6053) - implemented
- [Picture in Picture](https://github.com/organicmaps/organicmaps/issues/6650) - collecting feedback
- "another for dimming the screen when there are no upcoming turns (replicating the Waze behavior)." - please file this one with some examples of how it works.

**Comment by pastk** [comment]
> The proposed behavior is implemented by #6053.

Hmm, I think it breaks the use case of #196, because #6053 makes screen always on in navigation mode and I assume that was the "navigation on the bicycle" use case.

**Comment by pastk** [comment]
Looks like some users were using the following pattern:
- start the navigation
- turn off the screen
- turn on the screen when needed to e.g. check current location and whether its on the route still
- rely on screen going to sleep automatically

Still, its quite a niche use case. But just to remind, the "Allow screen to sleep" option was initially introduced for this use case (#196).

**Comment by pastk** [comment]
So what use cases do we have?

1. keep screen on for navigation
2. allow screen to sleep in navigation
3. keep screen on while browsing the map (planning, going somewhere w/o using navigation...)
4. allow screen to sleep to avoid battery draining (e.g. if a phone wakes up due to an accidental touch, etc.)

**Comment by biodranik** [comment]
Case (3) should not be on the list. If someone browses the map and then stops doing that, the screen should be automatically turned off, like in all other apps.
The missing cases are:
- Keep the screen on if the position is centered on the map not in the nav mode
- Maps are downloaded (until we finish a foreground service to download maps).

Point (4) is very controversial to me in this list. So far only one person complained to it (Roman), and consequences of changes arose after it was changed.

**Comment by pastk** [comment]
> Case (3) should not be on the list. If someone browses the map and then stops doing that, the screen should be automatically turned off, like in all other apps. 

Well, this is exactly my use case :) And this is exactly how Android version of OM (and I assume MM too) worked till #6053 (the screen was always on before the initial introduction of the setting too, see https://github.com/organicmaps/organicmaps/commit/470bc6f231bcb93a8024e8407d5b7903affdd9d2#diff-92baea9cd78cc337841445fde116fcd6b7e9dc6968d9cd56e855cbbf8a32b420).

It was convenient and configurable - better than the other apps! So let's not make it worse, please!

**Comment by biodranik** [comment]
Then the right solution is to introduce 3-state option:
- Keep the screen always on, in any situation
- Always sleep following the system setting (even in navigation mode)
- Automatic (default, do not sleep when position is centered or nav mode is active).

**Comment by pastk** [comment]
>     * Automatic (default, do not sleep when position is centered or nav mode is active).

Why make exception for the "centered" mode? And it'll be prone to the battery drain issue then.

IMO whether its centered or not is not related to whether a user is using OM still or just forgot to turn it off.
And its not easy to convey this logic to users. Many users just use the "center" button to jump to their current location and are not even aware or care that it'll "follow".

**Comment by biodranik** [comment]
@pastk dim and then sleep? Except when maps are downloaded, right?

**Comment by biodranik** [comment]
> IMO whether its centered or not is not related to whether a user is using OM still or just forgot to turn it off.

Of course, it is related. Many users use OM in a map-centered mode for free walking and driving. Locking the device in this situation is a bad idea. And that was the default behavior for 10 years, nobody complained. Then a specific use-case arose "I need to auto lock it even when nav/center mode is active, because it's inconvenient to lock the device manually on a bicycle". And a setting was added specifically for that case.

**Comment by pastk** [comment]
> Of course, it is related. Many users use OM in a map-centered mode for free walking and driving. Locking the device in this situation is a bad idea. And that was the default behavior for 10 years, nobody complained. 

1. Maybe it was the default behavior on iOS, I don't know. But on Android the default behavior was "screen always on" (regardless of centered / non-centered - the code history tells so, see my reference above and Romans' in #6053). Nobody complained :)
2. Why the "screen always on" (regardless of centered / non-centered) behavior won't work for this case?
3. Free walking is more convenient in non-centered mode
  - its very natural and easy to pan around to e.g. check what's further to the left
  - walking speed is low, hence automatic location following is not important compared to high-speed travel
  - even if you "lost yourself" its easy to go back to your position via one tap
  - but its not obvious or logical or convenient to center back just for the sake of keeping the screen on
  - pedestrians can easily spare time and their attention to browse map around and re-center when needed
    - as opposed to drivers and cyclists - for them the "centered/follow" mode is of much more use indeed, but see point 2.
    - also "free walking" is a common scenario, while "free driving / riding" is a rather niche case (most drivers and cyclists use navigation)

**Comment by denilsonsa** [comment]
> * "another for dimming the screen when there are no upcoming turns (replicating the Waze behavior)." - please file this one with some examples of how it works.

Someone requested it at #6045. I'm gonna share my experiences over there.

**Comment by biodranik** [comment]
Other important cases where the device should not be locked:

0. The current location is unknown.
1. OM is launched and starts to search for the current position, user does nothing and just stares at the screen until the position is found.
2. User presses the "find my position" button and waits for the position to be detected.

**Comment by pastk** [comment]
>     0. The current location is unknown.

This one is fine IMO. A user could've disabled the location or didn't give a permission or didn't confirm continuation of location search deliberately.

>     1. OM is launched and starts to search for the current position, user does nothing and just stares at the screen until the position is found.
> 

Not sure about this one as it could lead to the accidental battery drain (the phone wakes up accidentally and stays on indefinitely because location fix can not be obtained).

>     2. User presses the "find my position" button and waits for the position to be detected.

Right.

**Comment by biodranik** [comment]
>> 0. The current location is unknown.

> This one is fine IMO. A user could've disabled the location or didn't give a permission or didn't confirm continuation of location search deliberately.

1. If user disabled the location or didn't give the permission, then user will/should see a dialog after pressing location button. It's not related to the discussion.
2. The case when user pressed location button (or opened OM and started to find location) and then "forgot" about OM, expecting that the screen would turn off automatically, is unreal. The only possible situation here is when a device was randomly awakened.
3. @rtsisyk announced in #6053, and it looks like it doesn't work yet:
> This patch ... Forces the screen to keep on in PENDING_LOCATION no matter of the option.

---

> Not sure about this one as it could lead to the accidental battery drain (the phone wakes up accidentally and stays on indefinitely because location fix can not be obtained).

This is the default most-popular use-case of OM. If position was centered, then OM will try to show the map centered. If 30+ hours passed after not using the centered mode, OM assumes that you're on the go and you need your position centered again and does it.

We did not get any reports from users about this case in 12 years. An accidental launch of any app looks like a generic operating system issue, not like a maps-focused application issue.

Also it sounds weird: instead of focusing on improving the OM's UX, we're focusing on improving some random system bug while sacrificing the frequent/important app UX.

**Comment by pastk** [comment]
> > > 0. The current location is unknown.
> 
> > This one is fine IMO. A user could've disabled the location or didn't give a permission or didn't confirm continuation of location search deliberately.
> 
>     1. If user disabled the location or didn't give the permission, then user will/should see a dialog after pressing location button. It's not related to the discussion.
> 
>     2. The case when user pressed location button (or opened OM and started to find location) and then "forgot" about OM, expecting that the screen would turn off automatically, is unreal. The only possible situation here is when a device was randomly awakened.
> 

Yes, these are your case 3. from your post above (when a user tapped a location button).

I meant a case when the location is unknown and is not being searched for. E.g. I start OM, the location button is crossed out and I don't tap it because I don't want or care about my current location (e.g. exploring a map of the country far away).

**Comment by pastk** [comment]
Let's keep it open, we might adjust the behavior and/or introduce more options (like dim the screen during the navigation even) upon users' feedback.

**Comment by weckerle** [comment]
Is there already an issue for #6674 but for iOS? It would be great to have a switch in settings which ensures the app never locks the screen under any circumstances. Thanks!

**Comment by g1tc4t** [comment]
I want to bump this thread and combine it with [PR 437](https://github.com/organicmaps/organicmaps/pull/437) and another Idea: For power savings i'd like to have the screen being normally turned off but when there is a navigation instruction i'd like the screen to become awake so i can have a look at it to check the streets. The Screen should turn off again when the default screen-on-time configured in the OS elapsed.
I created a [Draft PR](https://github.com/organicmaps/organicmaps/pull/12920) and invite everybody to test this and give feedback.

**Comment by biodranik** [comment]
@g1tc4t thanks! What is the rationale for the PR? If a user wants to turn the screen off, then a power button press can do it, while still leaving voice commands active. Changing behavior because of the power saving settins would be very confusing for many users.

**Comment by Misalf-git** [comment]
> What is the rationale for the PR?

Apparently to save battery, especially when it's low, and to prevent fiddling with buttons while driving.

**Comment by biodranik** [comment]
Unexpected hiding of the important navigation screen may lead to accidents while driving.

**Comment by biodranik** [comment]
And most drivers have cables in cars to recharge their phones.

**Comment by Misalf-git** [comment]
> Unexpected hiding of the important navigation screen may lead to accidents while driving.

What!?

Seems like an AI response.

**Comment by biodranik** [comment]
That's not an AI response. I am, as a driver, would not appreciate the unexpectedly turned off screen when I need it to make a right turn. And the car may crash if I try to turn the screen back on after such an accident, because I will stop looking on the road.

**Comment by Misalf-git** [comment]
> And most drivers have cables in cars to recharge their phones.

I didn't know that. But not every vehicle (maybe car, bike, old camper or whatnot) necessarily provides the possibility to charge devices.

> I am, as a driver, would not appreciate the unexpectedly turned off screen when I need it to make a right turn.

Most drivers watch the road while making a turn.

Anyway, you might want to read the PR's description. It aims to let the display turn off _when it's **not** needed_.

This:
> And the car may crash if I try to turn the screen back on [...]

seems to contradict with this suggestion:
> If a user wants to turn the screen off, then a power button press can do it, [...]

It doesn't matter what action a button press would trigger, because, yes, it's potentially dangerous while driving.

**Comment by g1tc4t** [comment]
@biodranik you're totally right that the user could turn on/off the screen manually, but I think it would be quite convenient to have the screen turned on automatically when I need it most (show the correct turn instructions right before making the turn), and have it turned off to reduce power consumption when just driving along my way.
I use it for bike routing and I cannot charge the phone while riding, so keeping the screen off most of the time is not something unexpected.

Maybe it is not the best idea to couple an screen-auto-off-on function with the power savings settings, but this would void another user setting. And for me it is kind of a power savings setting.

**Comment by biodranik** [comment]
> but I think it would be quite convenient to have the screen turned on automatically when I need it most

Is it technically feasible? Can a locked/sleeping device be awakened programmatically again?

**Comment by g1tc4t** [comment]
Alarm apps are doing this, too. So why not?

**Comment by bukowa** [comment]
> [@biodranik](https://github.com/biodranik) you're totally right that the user could turn on/off the screen manually, but I think it would be quite convenient to have the screen turned on automatically when I need it most (show the correct turn instructions right before making the turn), and have it turned off to reduce power consumption when just driving along my way. I use it for bike routing and I cannot charge the phone while riding, so keeping the screen off most of the time is not something unexpected.
> 
> Maybe it is not the best idea to couple an screen-auto-off-on function with the power savings settings, but this would void another user setting. And for me it is kind of a power savings setting.

OsmAnd does it, it's ultra bugy but most of the time it works. Ultra battery saving, otherwise it just eats the battery on bike like it's nothing.

**Comment by bukowa** [comment]
https://github.com/osmandapp/OsmAnd/blob/82d43cd44b85eaf0da07d976a0fbefd6a80c6ab0/OsmAnd/src/net/osmand/plus/helpers/LockHelper.java
